fix staticcheck failures: apiserver/pkg/{storage,util}

obeyda opened the pull request:
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

fix static check failures for these folders:

- vendor/k8s.io/apiserver/pkg/storage
- vendor/k8s.io/apiserver/pkg/storage/cacher
- vendor/k8s.io/apiserver/pkg/storage/etcd3
- vendor/k8s.io/apiserver/pkg/storage/tests
- vendor/k8s.io/apiserver/pkg/storage/value/encrypt/envelope
- vendor/k8s.io/apiserver/pkg/util/webhook
- vendor/k8s.io/apiserver/pkg/util/wsstream 

Errors reported by static check:

```
+++ Running case: verify.staticcheck 
+++ working dir: /home/obeyda/go/src/k8s.io/kubernetes
+++ command: bash "hack/make-rules/../../hack/verify-staticcheck.sh"
Errors from staticcheck:
vendor/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:656:2: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
vendor/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:666:6: call to T.Fatalf
vendor/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:724:2: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
vendor/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:734:5: call to T.Fatalf
vendor/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:927:4: ineffective break statement. Did you mean to break out of the outer loop? (SA4011)
vendor/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:978:2: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
vendor/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:983:5: call to T.Fatalf
vendor/k8s.io/apiserver/pkg/storage/cacher/caching_object_test.go:137:3: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
vendor/k8s.io/apiserver/pkg/storage/cacher/caching_object_test.go:152:5: call to T.Fatalf
vendor/k8s.io/apiserver/pkg/storage/cacher/watch_cache_test.go:407:2: this value of err is never used (SA4006)
vendor/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:1914:3: field t is unused (U1000)
vendor/k8s.io/apiserver/pkg/storage/selection_predicate_test.go:33:6: type IgnoredList is unused (U1000)
vendor/k8s.io/apiserver/pkg/storage/tests/cacher_test.go:949:2: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
vendor/k8s.io/apiserver/pkg/storage/tests/cacher_test.go:959:6: call to T.Fatalf
vendor/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go:135:4: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
vendor/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go:141:6: call to T.Fatalf
vendor/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go:150:4: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
vendor/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go:159:6: call to T.Fatalf
vendor/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go:224:2: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
vendor/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go:231:4: call to T.Fatalf
vendor/k8s.io/apiserver/pkg/util/webhook/certs_test.go:71:5: var badCAKey is unused (U1000)
vendor/k8s.io/apiserver/pkg/util/wsstream/conn_test.go:58:2: the goroutine calls T.Fatal, which must be called in the same goroutine as the test (SA2002)
vendor/k8s.io/apiserver/pkg/util/wsstream/conn_test.go:62:4: call to T.Fatal
vendor/k8s.io/apiserver/pkg/util/wsstream/conn_test.go:75:2: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
vendor/k8s.io/apiserver/pkg/util/wsstream/conn_test.go:78:4: call to T.Fatalf
vendor/k8s.io/apiserver/pkg/util/wsstream/conn_test.go:140:2: the goroutine calls T.Fatal, which must be called in the same goroutine as the test (SA2002)
vendor/k8s.io/apiserver/pkg/util/wsstream/conn_test.go:144:4: call to T.Fatal
vendor/k8s.io/apiserver/pkg/util/wsstream/conn_test.go:157:2: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
vendor/k8s.io/apiserver/pkg/util/wsstream/conn_test.go:160:4: call to T.Fatalf

Please review the above warnings. You can test via:
  hack/verify-staticcheck.sh <failing package>
If the above warnings do not make sense, you can exempt the line or file. See:
  https://staticcheck.io/docs/#ignoring-problems

+++ exit code: 1
+++ error: 1
FAILED   verify-staticcheck.sh  12s
========================
FAILED TESTS
========================
hack/make-rules/../../hack/verify-staticcheck.sh
make: *** [Makefile:129: verify] Error 1
```

**Which issue(s) this PR fixes**:

Part of #92402

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

obeyda committed:
fix staticcheck failures: apiserver/pkg/{storage,util}

fix static check failures for these folders:
- vendor/k8s.io/apiserver/pkg/storage
- vendor/k8s.io/apiserver/pkg/storage/cacher
- vendor/k8s.io/apiserver/pkg/storage/etcd3
- vendor/k8s.io/apiserver/pkg/storage/tests
- vendor/k8s.io/apiserver/pkg/storage/value/encrypt/envelope
- vendor/k8s.io/apiserver/pkg/util/webhook
- vendor/k8s.io/apiserver/pkg/util/wsstream

Errors from staticcheck:
  vendor/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:656:2: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
  vendor/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:666:6: call to T.Fatalf
  vendor/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:724:2: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
  vendor/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:734:5: call to T.Fatalf
  vendor/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:927:4: ineffective break statement. Did you mean to break out of the outer loop? (SA4011)
  vendor/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:978:2: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
  vendor/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:983:5: call to T.Fatalf
  vendor/k8s.io/apiserver/pkg/storage/cacher/caching_object_test.go:137:3: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
  vendor/k8s.io/apiserver/pkg/storage/cacher/caching_object_test.go:152:5: call to T.Fatalf
  vendor/k8s.io/apiserver/pkg/storage/cacher/watch_cache_test.go:407:2: this value of err is never used (SA4006)
  vendor/k8s.io/apiserver/pkg/storage/etcd3/store_test.go:1914:3: field t is unused (U1000)
  vendor/k8s.io/apiserver/pkg/storage/selection_predicate_test.go:33:6: type IgnoredList is unused (U1000)
  vendor/k8s.io/apiserver/pkg/storage/tests/cacher_test.go:949:2: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
  vendor/k8s.io/apiserver/pkg/storage/tests/cacher_test.go:959:6: call to T.Fatalf
  vendor/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go:135:4: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
  vendor/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go:141:6: call to T.Fatalf
  vendor/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go:150:4: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
  vendor/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go:159:6: call to T.Fatalf
  vendor/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go:224:2: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
  vendor/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go:231:4: call to T.Fatalf
  vendor/k8s.io/apiserver/pkg/util/webhook/certs_test.go:71:5: var badCAKey is unused (U1000)
  vendor/k8s.io/apiserver/pkg/util/wsstream/conn_test.go:58:2: the goroutine calls T.Fatal, which must be called in the same goroutine as the test (SA2002)
  vendor/k8s.io/apiserver/pkg/util/wsstream/conn_test.go:62:4: call to T.Fatal
  vendor/k8s.io/apiserver/pkg/util/wsstream/conn_test.go:75:2: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
  vendor/k8s.io/apiserver/pkg/util/wsstream/conn_test.go:78:4: call to T.Fatalf
  vendor/k8s.io/apiserver/pkg/util/wsstream/conn_test.go:140:2: the goroutine calls T.Fatal, which must be called in the same goroutine as the test (SA2002)
  vendor/k8s.io/apiserver/pkg/util/wsstream/conn_test.go:144:4: call to T.Fatal
  vendor/k8s.io/apiserver/pkg/util/wsstream/conn_test.go:157:2: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
  vendor/k8s.io/apiserver/pkg/util/wsstream/conn_test.go:160:4: call to T.Fatalf

Part of #92402

obeyda commented:
/check-cla

MHBauer reviewed (commented):
/ok-to-test
Please squash the commits together, and maybe add the static check output to the commit message so we can understand why the changes are made.

fedebongio commented:
/assign @jingyih 
/triage accepted

obeyda commented:
Commits squashed, also I've updated the description with the static check output

jingyih commented:
Replacing fatal with error is probably not the proper way to fix the static check failure. To fix it, the err needs to be propagated back to the main test go routine. Something like:
```
func TestExample(t *testing.T) {
    errCh := make(chan error, 1)
    go func() {
        err := f()
        errCh <- err
    }()

    err := <-errCh
    if err != nil {
        t.Fatal(err)
    }
}
```

obeyda commented:
@jingyih  thanks for your review. PTAL

obeyda commented:
/retest

obeyda commented:
I think the failing tests aren't related to my changes, before rebase all tests were passing

obeyda commented:
/test pull-kubernetes-e2e-kind
/test pull-kubernetes-e2e-gce-ubuntu-containerd

obeyda commented:
@jingyih, PTAL

jingyih commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -682,9 +684,15 @@ func TestCacherNoLeakWithMultipleWatchers(t *testing.T) {
 			}
 		}
 	}()
+	go func() {
```
Sorry I do not follow why we need a new go routine here?

obeyda replied:
this is so we don't block and allow receive on errorCh on line 693

jingyih commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -731,7 +740,7 @@ func testCacherSendBookmarkEvents(t *testing.T, allowWatchBookmarks, expectedBoo
 				}})
 			if err != nil {
-				t.Fatalf("failed to add a pod: %v", err)
+				errorCh <- fmt.Errorf("failed to add a pod: %v", err)
```
Upon getting a non-nil error, we should return here? (as apposed to keep trying in the loop)

obeyda replied:
Agreed, will add it.
```suggestion
				errorCh <- fmt.Errorf("failed to add a pod: %v", err)
				return
```

jingyih commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -720,6 +728,7 @@ func testCacherSendBookmarkEvents(t *testing.T, allowWatchBookmarks, expectedBoo
 		t.Fatalf("Failed to create watch: %v", err)
 	}
 
+	errorCh := make(chan error, 1)
```
defer close this channel?

obeyda replied:
I didn't think we'll need to close this since it's used in a `select` statement, but if you think it's necessary I'll update that.

jingyih commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -924,7 +937,6 @@ func TestDispatchingBookmarkEventsWithConcurrentStop(t *testing.T) {
 
 		select {
 		case <-done:
-			break
```
Why this is removed?

obeyda replied:
because this is an ineffective break,
error from static check:
> ineffective break statement. Did you mean to break out of the outer loop? (SA4011)

jingyih commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -974,28 +986,45 @@ func TestBookmarksOnResourceVersionUpdates(t *testing.T) {
-			}
-			if event.Type == watch.Bookmark && rv == uint64(expectedRV) {
+			select {
+			case <-done:
```
Sorry I do not follow why we need the done channel to exit the function? Aren't the existing `return` statements in this function enough?

obeyda replied:
My bad 😄, this is actually not needed here (I had in mind this is multiple goroutines, but it's only one)

jingyih commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go:
```diff
@@ -974,28 +986,45 @@ func TestBookmarksOnResourceVersionUpdates(t *testing.T) {
 	cacher.watchCache.UpdateResourceVersion(strconv.Itoa(expectedRV))
 
-	wg.Wait()
+	go func() {
```
Same here, not sure why a new go routine is needed.

obeyda replied:
Same as above, to avoid blocking and allow receive on `if err = <- errorCh; err != nil {...}`

jingyih commented on staging/src/k8s.io/apiserver/pkg/storage/cacher/caching_object_test.go:
```diff
@@ -141,20 +143,34 @@ func TestCachingObjectRaces(t *testing.T) {
 				buffer.Reset()
 				if err := object.CacheEncode(encoder.identifier, encoder.encode, buffer); err != nil {
-					t.Errorf("unexpected error: %v", err)
+					errorCh <- fmt.Errorf("unexpected error: %v", err)
```
Is the staticcheck complaining these "Errorf" statements?

obeyda replied:
No, the staticcheck is complaining about the `Fatalf` that exists in this goroutine, but since I fixed that, I also updated the `Errorf` statements to use the same approach.

obeyda committed:
review: channel suggestions

remove unneeded channel
defer close for errorCh
add missing return after errorCh send
